cc @janvorli, @josalem, @AaronRobinsonMSFT

I am leaving out `Wplacement-new` for now, as those are bit weird to reason about, e.g.

```sh
[ 28%] Building CXX object vm/eventing/eventpipe/CMakeFiles/eventpipe_objs.dir/runtime/src/native/eventpipe/ep.c.o
In file included from /runtime/src/coreclr/classlibnative/../vm/threads.h:157,
                 from /runtime/src/coreclr/classlibnative/../vm/simplerwlock.hpp:10,
                 from /runtime/src/coreclr/classlibnative/../vm/stubmgr.h:49,
                 from /runtime/src/coreclr/classlibnative/../vm/clsload.hpp:20,
                 from /runtime/src/coreclr/classlibnative/../vm/common.h:350,
                 from /runtime/src/coreclr/classlibnative/bcltype/arraynative.cpp:12:
/runtime/src/coreclr/classlibnative/../vm/stackingallocator.h: In member function 'void* StackingAllocator::UnsafeAllocNoThrow(unsigned int)':
/runtime/src/coreclr/classlibnative/../vm/stackingallocator.h:160:52: error: placement new constructing an object of type 'Sentinal' and size '16' in a region of type 'char*' and size '0' [-Werror=placement-new=]
  160 |         m_FirstBlock->m_Sentinal = new(m_FirstFree - sizeof(Sentinal)) Sentinal(m_FirstBlock->m_Sentinal);
      |                                        ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
/runtime/src/coreclr/classlibnative/../vm/stackingallocator.h:200:23: note: at offset -16 from 'StackingAllocator::m_FirstFree' declared here
  200 |     char             *m_FirstFree;        // Pointer to first free byte in head block
      |                       ^~~~~~~~~~~
```

--- a/eng/native/configurecompiler.cmake
+++ b/eng/native/configurecompiler.cmake
@@ -399,8 +399,6 @@ if (CLR_CMAKE_HOST_UNIX)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-misleading-indentation>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-stringop-overflow>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-stringop-truncation>)
-    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-mismatched-new-delete>)
-    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-free-nonheap-object>)
     add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-Wno-placement-new>)
 
     check_cxx_compiler_flag(-faligned-new COMPILER_SUPPORTS_F_ALIGNED_NEW)

--- a/src/coreclr/vm/crst.h
+++ b/src/coreclr/vm/crst.h
@@ -421,6 +421,12 @@ public:
         return new BYTE[size];
     }
 
+    void operator delete(void* mem)
+    {
+        WRAPPER_NO_CONTRACT;
+        delete[] (BYTE*)mem;
+    }
+
 private:
     // Do not use inplace operator new on Crst. A wrong destructor would be called if the constructor fails.
     // Use CrstStatic or CrstExplicitInit instead of the inplace operator new.

